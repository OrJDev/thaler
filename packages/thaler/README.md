# thaler

> Isomorphic server-side functions

[![NPM](https://img.shields.io/npm/v/thaler.svg)](https://www.npmjs.com/package/thaler) [![JavaScript Style Guide](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)

## Install

```bash
npm i thaler
```

```bash
yarn add thaler
```

```bash
pnpm add thaler
```

## What?

`thaler` allows you to produce isomorphic functions that only runs on the server-side. This is usually ideal if you want to do server-side operations (e.g. database, files, etc.) on the client-side but without adding more abstractions such as defining extra REST endpoints, creating client-side utilities to communicate with the exact endpoint etc.

Another biggest benefit of this is that, not only it is great for isomorphic fullstack apps (i.e. metaframeworks like NextJS, SolidStart, etc.), if you're using TypeScript, type inference is also consistent too, so no need for extra work to manually wire-in types for both server and client.

## Functions

### `server$`

`server$` is the simplest of the `thaler` functions, it receives a callback for processing server [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) and returns a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response).

The returned function can then accept request options (which is the second parameter for the `Request` object), you can also check out [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/fetch)

```js
import { server$ } from 'thaler';

const getMessage = server$(async (request) => {
  const { greeting, receiver } = await request.json();

  return new Response(`${greeting}, ${receiver}!`, {
    status: 200,
  });
});

// Usage
const response = await getMessage({
  method: 'POST',
  body: JSON.stringify({
    greeting: 'Hello',
    receiver: 'World',
  }),
});

console.log(await response.text()); // Hello, World!
```

### `get$`

Similar to `server$` except that it can receive an object that can be converted into query params. The object can have a string or an array of strings as its values.

Only `get$` can accept search parameters and uses the `GET` method, which makes it great for creating server-side logic that utilizes caching.

```js
import { get$ } from 'thaler';

const getMessage = get$(async ({ greeting, receiver }) => {
  return new Response(`${greeting}, ${receiver}!`, {
    status: 200,
  });
});

// Usage
const response = await getMessage({
  greeting: 'Hello',
  receiver: 'World',
});

console.log(await response.text()); // Hello, World!
```

You can also pass some request configuration (same as `server$`) as the second parameter for the function, however `get$` cannot have `method` or `body`. The callback in `get$` can also receive the `Request` instance as the second parameter.

```js
import { get$ } from 'thaler';

const getUser = get$((search, request) => {
  // do stuff
});

const user = await getUser(search, {
  headers: {
    // do some header stuff
  },
});
```

### `post$`

If `get$` is for `GET`, `post$` is for `POST`. Instead of query parameters, the object it receives is converted into form data, so the object can accept not only a string or an array of strings, but also a [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob), a [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File), or an array of either of those types.

Only `post$` can accept form data and uses the `POST` method, which makes it great for creating server-side logic when building forms.

```js
import { post$ } from 'thaler';

const addMessage = post$(async ({ greeting, receiver }) => {
  await db.messages.insert({ greeting, receiver });
  return new Response(null, {
    status: 200,
  });
});

// Usage
await addMessage({
  greeting: 'Hello',
  receiver: 'World',
});
```

You can also pass some request configuration (same as `server$`) as the second parameter for the function, however `post$` cannot have `method` or `body`. The callback in `post$` can also receive the `Request` instance as the second parameter.

```js
import { post$ } from 'thaler';

const addMessage = post$((formData, request) => {
  // do stuff
});

await addMessage(formData, {
  headers: {
    // do some header stuff
  },
});
```

### `fn$`

Unlike `get$` and `post$`, `fn$` uses a superior form of serialization, so that not only it supports valid JSON values, it supports [an extended range of JS values](https://github.com/lxsmnsyc/seroval#supports).

```js
import { fn$ } from 'thaler';

const addUsers = fn$(async (users) => {
  const db = await import('./db');
  return Promise.all(users.map((user) => db.users.insert(user)));
});

await addUsers([
  { name: 'John Doe', email: 'john.doe@johndoe.com' },
  { name: 'Jane Doe', email: 'jane.doe@janedoe.com' },
]);
```

You can also pass some request configuration (same as `server$`) as the second parameter for the function, however `fn$` cannot have `method` or `body`. The callback in `fn$` can also receive the `Request` instance as the second parameter.

```js
import { fn$ } from 'thaler';

const addMessage = fn$((data, request) => {
  // do stuff
});

await addMessage(data, {
  headers: {
    // do some header stuff
  },
});
```

#### Scoping

Other functions can capture server-side scope but unlike the other functions, `fn$` has a special behavior: it can capture the client-side scope of where the function is declared on the client.

```js
import { fn$ } from 'thaler';

const prefix = 'Message:';

const getMessage = fn$(({ greeting, receiver }) => {
  // `prefix` is captured and sent to the server
  return `${prefix} "${greeting}, ${receiver}!"`;
});

console.log(await getMessage({ greeting: 'Hello', receiver: 'World' })); // Message: "Hello, World!"
```

> **Note**
> `fn$` can only capture local scope, and not global scope.

If you're using an imported value inside the callback, it's better to import it inside the callback in the form of dynamic `import`.

```js
// Don't do this
import { example } from './example';

fn$(() => {
  example();
});

// Do this
fn$(async () => {
  const { example } = await import('./example');
  example();
});
```

> **Warning**
> Be careful on capturing scopes, as the captured variables must only be the values that can be serialized by `fn$`. If you're using a value that can't be serialized inside the callback that is declared outside, it cannot be captured by `fn$` and will lead to runtime errors.

## Integrations

- [Vite](https://github.com/lxsmnsyc/thaler/tree/main/packages/vite)

## License

MIT Â© [lxsmnsyc](https://github.com/lxsmnsyc)
