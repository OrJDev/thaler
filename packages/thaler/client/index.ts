import seroval, { ServerValue } from 'seroval';
import ThalerError from '../shared/error';
import {
  ThalerPostInit,
  ThalerPostParam,
  ThalerFunctionInit,
  ThalerFunctions,
  ThalerFunctionTypes,
  ThalerGetInit,
  ThalerGetParam,
} from '../shared/types';
import {
  patchHeaders,
  serializeFunctionBody,
  toFormData,
  toURLSearchParams,
} from '../shared/utils';

type ServerHandlerRegistration = [type: 'server', id: string];
type GetHandlerRegistration = [type: 'get', id: string];
type PostHandlerRegistration = [type: 'post', id: string];
type FunctionHandlerRegistration = [type: 'fn', id: string];
type PureHandlerRegistration = [type: 'pure', id: string];

type HandlerRegistration =
  | ServerHandlerRegistration
  | GetHandlerRegistration
  | PostHandlerRegistration
  | FunctionHandlerRegistration
  | PureHandlerRegistration;

interface HandlerRegistrationResult {
  type: ThalerFunctionTypes;
  id: string;
}

export function $$register(
  ...[type, id]: HandlerRegistration
): HandlerRegistrationResult {
  return { type, id };
}

async function serverHandler(type: ThalerFunctionTypes, id: string, init: RequestInit) {
  patchHeaders(init, type);
  const result = await fetch(id, init);
  return result;
}

async function postHandler<P extends ThalerPostParam>(
  id: string,
  form: P,
  init: ThalerPostInit = {},
) {
  return serverHandler('post', id, {
    ...init,
    method: 'POST',
    body: toFormData(form),
  });
}

async function getHandler<P extends ThalerGetParam>(
  id: string,
  search: P,
  init: ThalerGetInit = {},
) {
  return serverHandler('get', `${id}?${toURLSearchParams(search).toString()}`, {
    ...init,
    method: 'GET',
  });
}

async function fnHandler<T extends ServerValue, R extends ServerValue>(
  id: string,
  scope: ServerValue[],
  value: T,
  init: ThalerFunctionInit = {},
): Promise<R> {
  const response = await serverHandler('fn', id, {
    ...init,
    method: 'POST',
    body: serializeFunctionBody({ scope, value }),
  });
  if (response.ok) {
    const serialized = await response.text();
    // eslint-disable-next-line no-eval
    return (0, eval)(serialized) as R;
  }
  throw new ThalerError(id);
}

async function pureHandler<T extends ServerValue, R extends ServerValue>(
  id: string,
  value: T,
  init: ThalerFunctionInit = {},
): Promise<R> {
  const response = await serverHandler('pure', id, {
    ...init,
    method: 'POST',
    body: seroval(value),
  });
  if (response.ok) {
    const serialized = await response.text();
    // eslint-disable-next-line no-eval
    return (0, eval)(serialized) as R;
  }
  throw new ThalerError(id);
}

export function $$clone(
  { type, id }: HandlerRegistrationResult,
  scope: ServerValue[],
): ThalerFunctions {
  switch (type) {
    case 'server':
      return Object.assign(serverHandler.bind(null, 'server', id), {
        type,
        id,
      });
    case 'post':
      return Object.assign(postHandler.bind(null, id), {
        type,
        id,
      });
    case 'get':
      return Object.assign(getHandler.bind(null, id), {
        type,
        id,
      });
    case 'fn':
      return Object.assign(fnHandler.bind(null, id, scope), {
        type,
        id,
      });
    case 'pure':
      return Object.assign(pureHandler.bind(null, id), {
        type,
        id,
      });
    default:
      throw new Error('unknown registration type');
  }
}
