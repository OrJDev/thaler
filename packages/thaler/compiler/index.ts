import * as babel from '@babel/core';
import path from 'path';
import thalerBabel, { PluginOptions } from './plugin';

export type Options = Omit<PluginOptions, 'source'>

export default async function compile(
  id: string,
  code: string,
  options: Options,
) {
  const pluginOption = [thalerBabel, {
    source: id,
    prefix: options.prefix,
    mode: options.mode,
  }];
  const plugins: NonNullable<NonNullable<babel.TransformOptions['parserOpts']>['plugins']> = ['jsx'];
  if (/\.[mc]?tsx?$/i.test(id)) {
    plugins.push('typescript');
  }
  const result = await babel.transformAsync(code, {
    plugins: [
      pluginOption,
    ],
    parserOpts: {
      plugins,
    },
    filename: path.basename(id),
    ast: false,
    sourceMaps: true,
    configFile: false,
    babelrc: false,
    sourceFileName: id,
  });

  if (result) {
    return {
      code: result.code || '',
      map: result.map,
    };
  }
  throw new Error('invariant');
}
