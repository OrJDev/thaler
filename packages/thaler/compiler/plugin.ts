import * as babel from '@babel/core';
import { addNamed } from '@babel/helper-module-imports';
import * as t from '@babel/types';
import { ThalerFunctionTypes } from '../shared/types';
import { getImportSpecifierKey, isPathValid } from './checks';
import getForeignBindings from './get-foreign-bindings';
import unwrapNode from './unwrap-node';
import xxHash32 from './xxhash32';

const IMPORTS = {
  register: '$$register',
  clone: '$$clone',
  scope: '$$scope',
};

interface TrackedImport {
  type: ThalerFunctionTypes;
  scoping: boolean;
}

const TRACKED_IMPORTS: Record<string, TrackedImport> = {
  server$: { type: 'server', scoping: false },
  post$: { type: 'post', scoping: false },
  get$: { type: 'get', scoping: false },
  fn$: { type: 'fn', scoping: true },
  pure$: { type: 'pure', scoping: false },
};

const SOURCE_MODULE = 'thaler';
const CLIENT_MODULE = 'thaler/client';
const SERVER_MODULE = 'thaler/server';

export interface PluginOptions {
  source: string;
  origin: string;
  prefix?: string;
  mode: 'server' | 'client';
}

interface State extends babel.PluginPass {
  imports: Map<string, t.Identifier>;
  registry: Map<t.Identifier, TrackedImport>;
  namespaces: Set<t.Identifier>;
  count: number;
  prefix: string;
  opts: PluginOptions;
}

function getImportIdentifier(
  state: State,
  path: babel.NodePath,
  name: string,
): t.Identifier {
  const current = state.imports.get(name);
  if (current) {
    return current;
  }
  const target = state.opts.mode === 'server'
    ? SERVER_MODULE
    : CLIENT_MODULE;
  const id = addNamed(path, name, target);
  state.imports.set(name, id);
  return id;
}

function extractImportIdentifiers(
  ctx: State,
  path: babel.NodePath<t.ImportDeclaration>,
) {
  const mod = path.node.source.value;

  // Identify hooks
  if (mod === SOURCE_MODULE) {
    for (let i = 0, len = path.node.specifiers.length; i < len; i++) {
      const specifier = path.node.specifiers[i];
      switch (specifier.type) {
        case 'ImportSpecifier': {
          const key = getImportSpecifierKey(specifier);
          if (key in TRACKED_IMPORTS) {
            ctx.registry.set(specifier.local, TRACKED_IMPORTS[key]);
          }
        }
          break;
        case 'ImportNamespaceSpecifier':
          ctx.namespaces.add(specifier.local);
          break;
        default:
          break;
      }
    }
  }
}

function getRootStatementPath(path: babel.NodePath) {
  let current = path.parentPath;
  while (current) {
    const next = current.parentPath;
    if (t.isProgram(next)) {
      return current;
    }
    current = next;
  }
  return path;
}

function createThalerFunction(
  ctx: State,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
  registry: TrackedImport,
) {
  const argument = path.get('arguments')[0];
  if (
    argument
    && (
      isPathValid(argument, t.isArrowFunctionExpression)
      || isPathValid(argument, t.isFunctionExpression)
    )
  ) {
    // Create an ID
    const id = `${ctx.prefix}${ctx.count}`;
    ctx.count += 1;
    // Create the call expression
    const args: t.Expression[] = [t.stringLiteral(registry.type), t.stringLiteral(id)];
    if (ctx.opts.mode === 'server') {
      // Hoist the argument
      args.push(argument.node);
    }

    // Create registration call
    const registerID = path.scope.generateUidIdentifier(registry.type);
    const register = t.callExpression(
      getImportIdentifier(ctx, path, IMPORTS.register),
      args,
    );
    // Locate root statement (the top-level statement)
    const rootStatement = getRootStatementPath(path);
    // Push the declaration
    rootStatement.insertBefore(
      t.variableDeclaration(
        'const',
        [t.variableDeclarator(registerID, register)],
      ),
    );
    // Setup for clone call
    const cloneArgs: t.Expression[] = [registerID];
    // Collect bindings for scoping
    if (registry.scoping) {
      const scope = getForeignBindings(argument);
      cloneArgs.push(t.arrowFunctionExpression([], t.arrayExpression(scope)));
      if (scope.length) {
        // Add scoping to the arrow function
        if (ctx.opts.mode === 'server') {
          const statement = t.isStatement(argument.node.body)
            ? argument.node.body
            : t.blockStatement([
              t.returnStatement(argument.node.body),
            ]);
          statement.body = [
            t.variableDeclaration(
              'const',
              [
                t.variableDeclarator(
                  t.arrayPattern(scope),
                  t.callExpression(getImportIdentifier(ctx, path, IMPORTS.scope), []),
                ),
              ],
            ),
            ...statement.body,
          ];
        }
      }
    }
    // Replace with clone
    path.replaceWith(
      t.callExpression(
        getImportIdentifier(ctx, path, IMPORTS.clone),
        cloneArgs,
      ),
    );
  }
}

function transformCall(
  ctx: State,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
) {
  const trueID = unwrapNode(path.node.callee, t.isIdentifier);
  if (trueID) {
    const binding = path.scope.getBindingIdentifier(trueID.name);
    if (binding) {
      const registry = ctx.registry.get(binding);
      if (registry) {
        createThalerFunction(ctx, path, registry);
      }
    }
  }
  const trueMemberExpr = unwrapNode(path.node.callee, t.isMemberExpression);
  if (
    trueMemberExpr
    && !trueMemberExpr.computed
    && t.isIdentifier(trueMemberExpr.property)
  ) {
    const obj = unwrapNode(trueMemberExpr.object, t.isIdentifier);
    if (obj) {
      const binding = path.scope.getBindingIdentifier(obj.name);
      if (binding && ctx.namespaces.has(binding)) {
        const property = TRACKED_IMPORTS[trueMemberExpr.property.name];
        if (property) {
          createThalerFunction(ctx, path, property);
        }
      }
    }
  }
}

const DEFAULT_PREFIX = '__thaler';

function getPrefix(ctx: State) {
  const prefix = ctx.opts.prefix == null ? DEFAULT_PREFIX : ctx.opts.prefix;
  let file = '';
  if (ctx.opts.source) {
    file = ctx.opts.source;
  } else if (ctx.filename) {
    file = ctx.filename;
  }
  return `${ctx.opts.origin}/${prefix}/${xxHash32(file).toString(16)}-`;
}

export default function thalerPlugin(): babel.PluginObj<State> {
  return {
    name: 'thaler',
    pre() {
      this.imports = new Map();
      this.registry = new Map();
      this.namespaces = new Set();
      this.count = 0;
    },
    visitor: {
      Program(programPath, ctx) {
        ctx.prefix = getPrefix(ctx);
        programPath.traverse({
          ImportDeclaration(path) {
            extractImportIdentifiers(ctx, path);
          },
        });
      },
      CallExpression(path, ctx) {
        transformCall(ctx, path);
      },
      OptionalCallExpression(path, ctx) {
        transformCall(ctx, path);
      },
    },
  };
}
