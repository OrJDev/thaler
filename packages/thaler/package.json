{
  "name": "thaler",
  "version": "0.5.1",
  "type": "module",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=10"
  },
  "license": "MIT",
  "keywords": [
    "pridepack"
  ],
  "devDependencies": {
    "@types/babel__core": "^7.20.0",
    "@types/babel__traverse": "^7.18.3",
    "@types/node": "^18.13.0",
    "eslint": "^8.37.0",
    "eslint-config-lxsmnsyc": "^0.5.1",
    "pridepack": "2.4.1",
    "tslib": "^2.4.1",
    "typescript": "^4.9.4",
    "vitest": "^0.29.1"
  },
  "dependencies": {
    "@babel/core": "^7.21.0",
    "@babel/helper-module-imports": "^7.18.6",
    "@babel/traverse": "^7.21.2",
    "@babel/types": "^7.21.2",
    "seroval": "^0.7.0"
  },
  "scripts": {
    "prepublishOnly": "pridepack clean && pridepack build",
    "build": "pridepack build",
    "type-check": "pridepack check",
    "lint": "pridepack lint",
    "clean": "pridepack clean",
    "watch": "pridepack watch",
    "start": "pridepack start",
    "dev": "pridepack dev",
    "test": "vitest"
  },
  "description": "Isomorphic server-side functions",
  "repository": {
    "url": "https://github.com/lxsmnsyc/thaler.git",
    "type": "git"
  },
  "homepage": "https://github.com/lxsmnsyc/thaler/tree/main/packages/thaler",
  "bugs": {
    "url": "https://github.com/lxsmnsyc/thaler/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Alexis Munsayac",
  "private": false,
  "typesVersions": {
    "*": {
      "compiler": [
        "./dist/types/compiler/index.d.ts"
      ],
      "client": [
        "./dist/types/client/index.d.ts"
      ],
      "server": [
        "./dist/types/server/index.d.ts"
      ],
      "utils": [
        "./dist/types/utils/index.d.ts"
      ]
    }
  },
  "types": "./dist/types/src/index.d.ts",
  "main": "./dist/cjs/production/index.cjs",
  "module": "./dist/esm/production/index.mjs",
  "exports": {
    ".": {
      "development": {
        "require": "./dist/cjs/development/index.cjs",
        "import": "./dist/esm/development/index.mjs"
      },
      "require": "./dist/cjs/production/index.cjs",
      "import": "./dist/esm/production/index.mjs",
      "types": "./dist/types/src/index.d.ts"
    },
    "./compiler": {
      "development": {
        "require": "./dist/cjs/development/compiler.cjs",
        "import": "./dist/esm/development/compiler.mjs"
      },
      "require": "./dist/cjs/production/compiler.cjs",
      "import": "./dist/esm/production/compiler.mjs",
      "types": "./dist/types/compiler/index.d.ts"
    },
    "./client": {
      "development": {
        "require": "./dist/cjs/development/client.cjs",
        "import": "./dist/esm/development/client.mjs"
      },
      "require": "./dist/cjs/production/client.cjs",
      "import": "./dist/esm/production/client.mjs",
      "types": "./dist/types/client/index.d.ts"
    },
    "./server": {
      "development": {
        "require": "./dist/cjs/development/server.cjs",
        "import": "./dist/esm/development/server.mjs"
      },
      "require": "./dist/cjs/production/server.cjs",
      "import": "./dist/esm/production/server.mjs",
      "types": "./dist/types/server/index.d.ts"
    },
    "./utils": {
      "development": {
        "require": "./dist/cjs/development/utils.cjs",
        "import": "./dist/esm/development/utils.mjs"
      },
      "require": "./dist/cjs/production/utils.cjs",
      "import": "./dist/esm/production/utils.mjs",
      "types": "./dist/types/utils/index.d.ts"
    }
  }
}
