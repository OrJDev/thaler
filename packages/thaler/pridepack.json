{
  "target": "es2018",
  "entrypoints": {
    ".": "src/index.ts",
    "./babel": "./babel/index.ts",
    "./client": "./client/index.ts",
    "./server": "./server/index.ts"
  }
}