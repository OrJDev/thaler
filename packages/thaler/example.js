import compile from 'thaler/compiler';

const serverOptions = {
  prefix: 'api/__thaler',
  mode: 'server',
};

const FILE = 'src/index.ts';

const code = `
import { createResource, createSignal, Suspense } from 'solid-js';
import { fn$ } from 'thaler';
import { debounce } from 'thaler/utils';

const sleep = (ms: number) => new Promise((res) => {
  setTimeout(res, ms, true);
});

export default function Example() {
  const [state, setState] = createSignal(0);

  const prefix = 'Server Count';

  const serverCount = debounce(fn$(async (value: number) => {
    await sleep(1000);
    console.log('Received', value);
    return \`\${prefix}: \${value}\`;
  }), {
    key: () => 'sleep',
  });
  const [data] = createResource(state, (value) => serverCount(value));

  function increment() {
    setState((c) => c + 1);
  }

  return (
    <>
      <button onClick={increment}>
        {\`Client Count: \${state()}\`}
      </button>
      <div>
        <Suspense fallback={<h1>Loading</h1>}>
          <h1>{data()}</h1>
        </Suspense>
      </div>
    </>
  );
}


`;

console.log((await compile(FILE, code, serverOptions)).code);