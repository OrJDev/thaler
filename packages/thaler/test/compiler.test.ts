import * as babel from '@babel/core';
import { describe, expect, it } from 'vitest';
import plugin, { Options } from '../babel';

const options: Options = {
  origin: 'http://localhost:3000',
  prefix: 'example',
  source: 'example.tsx',
  mode: 'server',
};

async function compile(mode: 'server' | 'client', code: string) {
  const result = await babel.transformAsync(code, {
    plugins: [
      [plugin, { ...options, mode }],
    ],
    parserOpts: {
      plugins: [
        'jsx',
      ],
    },
  });

  return result?.code ?? '';
}

describe('server$', () => {
  it('should transform', async () => {
    const code = `
import { server$ } from 'thaler';

const example = server$((request) => {
  return new Response('Hello World', {
    headers: {
      'content-type': 'text/html',
    },
    status: 200,
  });
});
  `;
    expect(await compile('server', code)).toMatchSnapshot();
    expect(await compile('client', code)).toMatchSnapshot();
  });
});
describe('loader$', () => {
  it('should transform', async () => {
    const code = `
import { loader$ } from 'thaler';

const example = loader$(({ greeting, receiver}) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html',
    },
    status: 200,
  });
});
  `;
    expect(await compile('server', code)).toMatchSnapshot();
    expect(await compile('client', code)).toMatchSnapshot();
  });
});
describe('action$', () => {
  it('should transform', async () => {
    const code = `
import { action$ } from 'thaler';

const example = action$(({ greeting, receiver }) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html',
    },
    status: 200,
  });
});
  `;
    expect(await compile('server', code)).toMatchSnapshot();
    expect(await compile('client', code)).toMatchSnapshot();
  });
});
describe('fn$', () => {
  it('should transform', async () => {
    const code = `
import { fn$ } from 'thaler';

const PREFIX = 'Message: ';

const example = fn$(({ greeting, receiver }) => {
  const message = PREFIX + greeting + ', ' + receiver + '!';
  return message;
});
  `;
    expect(await compile('server', code)).toMatchSnapshot();
    expect(await compile('client', code)).toMatchSnapshot();
  });
});

describe('pure$', () => {
  it('should transform', async () => {
    const code = `
import { pure$ } from 'thaler';

const sleep = (ms) => new Promise((res) => {
  setTimeout(res, ms, true);
});

const example = pure$(async ({ greeting, receiver }) => {
  await sleep(1000);
  const message = greeting + ', ' + receiver + '!';
  return message;
});
  `;
    expect(await compile('server', code)).toMatchSnapshot();
    expect(await compile('client', code)).toMatchSnapshot();
  });
});
