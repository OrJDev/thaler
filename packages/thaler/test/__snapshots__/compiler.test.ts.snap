// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fn$ > should transform 1`] = `
"import { $$clone as _$$clone } from \\"thaler/server\\";
import { $$scope as _$$scope } from \\"thaler/server\\";
import { $$register as _$$register } from \\"thaler/server\\";
import { fn$ } from 'thaler';
const PREFIX = 'Message: ';
const _fn = _$$register(\\"fn\\", \\"http://localhost:3000/example/a4dc1948-0\\", ({
  greeting,
  receiver
}) => {
  const [PREFIX] = _$$scope();
  const message = PREFIX + greeting + ', ' + receiver + '!';
  return message;
});
const example = _$$clone(_fn, [PREFIX]);"
`;

exports[`fn$ > should transform 2`] = `
"import { $$clone as _$$clone } from \\"thaler/client\\";
import { $$register as _$$register } from \\"thaler/client\\";
import { fn$ } from 'thaler';
const PREFIX = 'Message: ';
const _fn = _$$register(\\"fn\\", \\"http://localhost:3000/example/a4dc1948-0\\");
const example = _$$clone(_fn, [PREFIX]);"
`;

exports[`get$ > should transform 1`] = `
"import { $$clone as _$$clone } from \\"thaler/server\\";
import { $$register as _$$register } from \\"thaler/server\\";
import { get$ } from 'thaler';
const _get = _$$register(\\"get\\", \\"http://localhost:3000/example/a4dc1948-0\\", ({
  greeting,
  receiver
}) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html'
    },
    status: 200
  });
});
const example = _$$clone(_get);"
`;

exports[`get$ > should transform 2`] = `
"import { $$clone as _$$clone } from \\"thaler/client\\";
import { $$register as _$$register } from \\"thaler/client\\";
import { get$ } from 'thaler';
const _get = _$$register(\\"get\\", \\"http://localhost:3000/example/a4dc1948-0\\");
const example = _$$clone(_get);"
`;

exports[`post$ > should transform 1`] = `
"import { $$clone as _$$clone } from \\"thaler/server\\";
import { $$register as _$$register } from \\"thaler/server\\";
import { post$ } from 'thaler';
const _post = _$$register(\\"post\\", \\"http://localhost:3000/example/a4dc1948-0\\", ({
  greeting,
  receiver
}) => {
  const message = greeting + ', ' + receiver + '!';
  return new Response(message, {
    headers: {
      'content-type': 'text/html'
    },
    status: 200
  });
});
const example = _$$clone(_post);"
`;

exports[`post$ > should transform 2`] = `
"import { $$clone as _$$clone } from \\"thaler/client\\";
import { $$register as _$$register } from \\"thaler/client\\";
import { post$ } from 'thaler';
const _post = _$$register(\\"post\\", \\"http://localhost:3000/example/a4dc1948-0\\");
const example = _$$clone(_post);"
`;

exports[`pure$ > should transform 1`] = `
"import { $$clone as _$$clone } from \\"thaler/server\\";
import { $$register as _$$register } from \\"thaler/server\\";
import { pure$ } from 'thaler';
const sleep = ms => new Promise(res => {
  setTimeout(res, ms, true);
});
const _pure = _$$register(\\"pure\\", \\"http://localhost:3000/example/a4dc1948-0\\", async ({
  greeting,
  receiver
}) => {
  await sleep(1000);
  const message = greeting + ', ' + receiver + '!';
  return message;
});
const example = _$$clone(_pure);"
`;

exports[`pure$ > should transform 2`] = `
"import { $$clone as _$$clone } from \\"thaler/client\\";
import { $$register as _$$register } from \\"thaler/client\\";
import { pure$ } from 'thaler';
const sleep = ms => new Promise(res => {
  setTimeout(res, ms, true);
});
const _pure = _$$register(\\"pure\\", \\"http://localhost:3000/example/a4dc1948-0\\");
const example = _$$clone(_pure);"
`;

exports[`server$ > should transform 1`] = `
"import { $$clone as _$$clone } from \\"thaler/server\\";
import { $$register as _$$register } from \\"thaler/server\\";
import { server$ } from 'thaler';
const _server = _$$register(\\"server\\", \\"http://localhost:3000/example/a4dc1948-0\\", request => {
  return new Response('Hello World', {
    headers: {
      'content-type': 'text/html'
    },
    status: 200
  });
});
const example = _$$clone(_server);"
`;

exports[`server$ > should transform 2`] = `
"import { $$clone as _$$clone } from \\"thaler/client\\";
import { $$register as _$$register } from \\"thaler/client\\";
import { server$ } from 'thaler';
const _server = _$$register(\\"server\\", \\"http://localhost:3000/example/a4dc1948-0\\");
const example = _$$clone(_server);"
`;
